{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# July 10\n",
    "\n",
    "### Assembly\n",
    "\n",
    "An assembly is a heirarchical data structure that maps the sequence data to a putative reconstruction of the target\n",
    "\n",
    "* Only way to assemble chromosomes or relate your assembly back to specific chromosomes is to do genetic crossing\n",
    "    * Generally, we will be talking about scaffold-level sequencing\n",
    "    \n",
    "Contigs: contiguous, ungapped seuqnce\n",
    "* Groups reads together\n",
    "* provides a multiple sequence alignment\n",
    "* Provides a consensus sequence\n",
    "\n",
    "Scaffolds: supercontigs\n",
    "* Defines contig order and orientation\n",
    "* defines size of gaps between contigs\n",
    "\n",
    "### Sequence reads\n",
    "* Sanger sequencing: 600-800bp\n",
    "    * Use fluorescent, chain terminating nucleotides bind to a DNA strand, then you can run a gel and check the sequence as it's being built up\n",
    "    * High quality reads -> low error rate\n",
    "\n",
    "\n",
    "* Second generation\n",
    "    * Roche 454\n",
    "        * Same kind of concept as Sanger, but uses beads to amplify DNA, then reads from beads\n",
    "    * Early Illumina (2007-2008)\n",
    "        * 36bp reads orginally, high error rate, but could get 10M reads in one lane\n",
    "    * Illumina\n",
    "        * 100bp, 150bp, 250bp depending on the machine, good quality reads\n",
    "        * 400M reads per lane\n",
    "        * uses flow cells, clonal bridge amplification, then elongate with fluorescent primers and read colors\n",
    "    \n",
    "    * Illumina reads\n",
    "        * Insert length of ~600bp\n",
    "        * Paired end sequencing reads 100bp from both sides\n",
    "        * allows you to possibly find gaps between scaffolds if you need to\n",
    "        \n",
    "        * Mate-pair reads\n",
    "            * 2-10k DNA molecule (2kb, 5kb, 10kb), circularized with Biotin\n",
    "            * Sonicate,then sequence the junction. Can use it to find larger gaps\n",
    "            * Used to arrange contigs\n",
    "            * 40kb mate-pairs by circularizing and shearing fosmids\n",
    "            * 150-200kb mate-pairs by circularizing and shearing BACs\n",
    "            \n",
    "            \n",
    "* Third generation\n",
    "    * Synthetic single molecule sequencing\n",
    "        * Moleculo - 10kb median length\n",
    "        * 10x Genomics - sparse, full molecule sequencing\n",
    "            * Shear long pieces of DNA and suspend in oil droplets, each droplet gets its own barcode, samples from the same initial piece of DNA should have barcodes close in number \n",
    "    \n",
    "    * True single molecule sequencing\n",
    "        * PacBio - 15kb median length, up to 25kb-50kb (read lengths are bound by length of initial DNA)\n",
    "            * Error rate ~ 15%, truly random error rate (vs sequence specific)\n",
    "        * Nanopore - 100kb???\n",
    "            * error rate >> 15%\n",
    "            * maybe can be used for final scaffolding since reads are so long\n",
    "\n",
    "\n",
    "* Assembly Concepts\n",
    "    * Overlap / Layout / Consensus\n",
    "        * At it's most basic, just overlay each gene fragment and find the matching sequences, where the sequences are the nodes and links are the edges\n",
    "        * **Hamiltonian cycle: visit each vertex one time**\n",
    "        \n",
    "        1. Overlaps must be discovered by an all-versus-all read comparison\n",
    "            * Seed and extend algorithm is used: k-mer matching followed by alignment extension\n",
    "            * Big O - O(n^2) runtime. Have to compare everything to everything else, runtime increases exponentially\n",
    "        2. Contruct the overlap graph\n",
    "            * Reads are nodes, overlaps are edges\n",
    "            * Must deal with cycles and tangles in the graph\n",
    "        3. The precise layout and consensus sequence is determined by doing a multiple sequence alignment\n",
    "            * Progressive pair-wise alignments\n",
    "            \n",
    "        Newbler:\n",
    "            * Construct unitigs: uncontested subgraphs\n",
    "            * Seed a second overlap graph with unitigs to construct contigs\n",
    "        \n",
    "        * Because every read is a node, the number of nodes scales with the number of reads\n",
    "        * Overlap computations become intractable on millions of reads\n",
    "        * BUT, if read lengths increase substntially, OLC may once again be preferable\n",
    "            * With PacBio, it is\n",
    "            \n",
    "    * Sequencing by Hybridization\n",
    "        * A sequencing chip containing all possible, say, 3-mers (4^3 = 64 3-mers). Wash DNA across chip, DNA will hybridize to 3-mers contained within its sequence. Look at fluorescence to assemble 3-mers back into full sequence\n",
    "        \n",
    "        * K-mers, a string with length of k\n",
    "            * Numbers of k-mers in a string = strlen - k + 1\n",
    "            * Number of possible k-mers: 4^k. As k approaches 19, approaches unique sequences\n",
    "            \n",
    "     * De Bruijn graphs\n",
    "         * Instead of each read representing a node in the graph, we will now let each unique sequence in the genome be represented by a node.\n",
    "         * We can represent unique sequence with k-mers\n",
    "         * Vertices (nodes) are (k-1)-mers, edges are k-mers\n",
    "         * **A Eulerian path: visit each edge once**\n",
    "         \n",
    "         * Assembly is a byproduct of graph construction\n",
    "         * Construct k-mers from reads, each read induces a path\n",
    "         * K-mer hashing finds all k-mer overlaps\n",
    "             * Ask if you've seen a k-mer before, if you haven't start a path, if you have, go to that path and continue\n",
    "             \n",
    "         * If:\n",
    "             1. All k-mers are present in the genome\n",
    "             2. All k-mers are error free\n",
    "             3. Each k-mer appears exactly once\n",
    "             (None of these will be true)\n",
    "         * Then, a Eulerian path can be found and the genome can be perfectly reconstructed from the path\n",
    "         \n",
    "         \n",
    "* Sequencing error\n",
    "    * One sequencing error induces up to K errorneous k-mers\n",
    "    * \"Spurs\" or \"tips\" are introduced (in De Bruijn) by sequencing errors, assemblers will find them and cut them out\n",
    "    * Bubbles are induced by sequencing error or polymorphism\n",
    "    * Repeat sequences cause path convergence (a repeat with 41 k-mers will give same 41 k-mers each time, they will be collapsed down\n",
    "    * Cycles are introduced by short repeats\n",
    "    \n",
    "    * Bandage for De Bruijn graph visualization\n",
    "    \n",
    "  \n",
    "* Sequencing Strategies\n",
    "    * Not really proper to just sequence then find an appropriate assembly method after the fact\n",
    "    * Algorithms tuned to assembly certain types of sequencing\n",
    "    \n",
    "    1. ALLPATHS-LG\n",
    "        * 2x100bp Illumina reads\n",
    "            * 180bp insert length (Overlapping ends)\n",
    "            * 800bp insert length\n",
    "            * 2kb, 5kb, 10kb mate pairs\n",
    "            * 20x PacBio to try to join scaffolds\n",
    "     \n",
    "     2. Discovar *de novo* Strategy\n",
    "         * 2x250bp Illumina Reads\n",
    "             * 800bp insert length\n",
    "             * 2kb, 5kb, 10kb mate pairs\n",
    "             * 20kb PacBio\n",
    "         * JGI's Meraculous assembler has been adapted to this\n",
    "         \n",
    "     3. BGI Strategy\n",
    "         * SOAP *de novo*\n",
    "         * Many PE insert lengths\n",
    "         \n",
    "     4. 10X Strategy\n",
    "         * 1 10X Illumina 2x150bp\n",
    "         \n",
    "     5. Pure PacBio Strategy\n",
    "         * 60x PacBio coverage"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# July 11\n",
    "### Short-read alignment\n",
    "\n",
    "* Gene expression quantification\n",
    "    * reads align to exons, have to use annotation file to assemble those back to the exons to check expression\n",
    "    * Aligners have to make sure reads spanning two or more exons are not penalized\n",
    "\n",
    "\n",
    "* Structural Variation Detection\n",
    "    * Use DNA to determine structural variation of chromosomes\n",
    "    * Duplications lead to increased read coverage for certain regions compared to control genomes\n",
    "    * Inversions look like they have flipped reads that map relatively far away compared to control\n",
    "\n",
    "\n",
    "* Metagenomics\n",
    "    * Programs take database of all organisms, then compares all DNA in a sample (e.g. of gut bacteria or sea bacteria across the world) to that database to find which organisms are in that sample and in what proportion\n",
    "    \n",
    "\n",
    "* Assembly verification\n",
    "    * Map all your reads back to your assembled genome to make sure your scaffold sequences came from *somewhere*.\n",
    "    \n",
    "    \n",
    "* Medical exome sequencing\n",
    "    * Take exons and generate complementary sequences for those exons\n",
    "    * Use RNA bait to pull out those exon sequences, then sequence and compare back to the reference genome to find where there might have been disease-causing mutations\n",
    "    \n",
    "    \n",
    "* BLAST: Optimal Alignment\n",
    "    * Smith-Waterman alignment, dynamic programming\n",
    "    1. BLAST provides an optimal alignment using dynamic program\n",
    "    2. BLAST was designed for searching for distantly related homologues\n",
    "    3. BLAST is optimized for gene-sized lengths of sequence\n",
    "    4. Computational complexity: O(m\\*n)\n",
    "    5. When **n** is a reference genome, the problem is not tractable\n",
    "    \n",
    "    \n",
    "* Alignment Algorithms\n",
    "    * BLAST and BLAT vs. Short-read aligners\n",
    "    \n",
    "        1. Must be capable of aligning hundreds of millions of reads\n",
    "            * Extremely fast -- less sensitive\n",
    "        2. Requires indexing and storing the reference genome in a special data structure\n",
    "        3. Can typically be multithreaded to increase efficiency\n",
    "        \n",
    "    * Types of alignments\n",
    "        1. Placement alignments\n",
    "            * Only the best placement needs to be calculated\n",
    "            * Reads will be fully aligned by additional software once they are in the correct location\n",
    "            * Bowtie, BWA\n",
    "            * Exome sequencing for disease detection\n",
    "                * Can be hard to distinguish real de novo mutations from sequencing error\n",
    "        2. Full alignments\n",
    "            * Place and fully align the read in one step\n",
    "            * GSnap\n",
    "            * RNA-seq, RAD-seq alignments\n",
    "            \n",
    "    * K-mer hashing algorithm\n",
    "        * GSnap\n",
    "        * More flexible, perhaps too promiscuous\n",
    "        * Can handle intron/exon splicing directly\n",
    "        * BLAST matches k-mers to reference, then performed non-gapped extension on those k-mer matches\n",
    "        * BLAT - BLAST-Like Alignment Tool\n",
    "            * Three major differences from BLAST:\n",
    "                1. BLAST builds an index of the query sequence and scans linearly through the BLAST database; BLAT indexes the database and scans linearly through the query sequence\n",
    "                2. BLAST returns each area of homology as a separate alignment; BLAT stitches alignments together.\n",
    "                3. BLAT has unique code for intron handling \"unsplices\" mRNA onto the genome\n",
    "                \n",
    "    * Burrows-Wheeler data structure\n",
    "        * used in data compression, e.g. bzip2\n",
    "        * Michael Burrows and David Wheeler in 1994 DEC\n",
    "        * Bowtie, BWA.\n",
    "        * Extremely compact and fast\n",
    "        * (-) Requires very close matches to the genome\n",
    "        * (-) to handle exon/intron splicing, must be coupled with an additional program\n",
    "        * \"magicable\"?\n",
    "        \n",
    "        * Encode: \"aardvark\"\n",
    "        * Rotate the text, add a sentinel character (a character that you don't expect to appear in your word)\n",
    "        * aardvark\\$, ardvark\\$a, rdvark\\$aa... \\$aardvark -> makes all possible suffices and prefices of word\n",
    "        * Sort the text, then extract last column -> \"k\\$avrraad\". Can now be encoded using run-length encoding: \"k\\$av2r2ad\" (compress the text)\n",
    "        * ???? wat\n",
    "        * The Last First Mapping Property\n",
    "            * The letter in the final column is previous to the letter in the first column.\n",
    "            * The relative, sorted order of letters in the final column is maintained in the first column\n",
    "            \n",
    "        * Bowtie - uses BWT\n",
    "                \n",
    "    * Suffix Trees\n",
    "        * Take a sequence, like banana, and do the same transformation banana\\$, This time, only keep the suffixes\n",
    "        * banana\\$, anana\\$, nana\\$ ... a\\$, \\$\n",
    "        \n",
    "    * Suffix array\n",
    "        * Take suffixes as in suffix trees, then put into array and sort the array\n",
    "        * Use a binary search to find your sequence (ideally O(n) complexity)\n",
    "        \n",
    "        \n",
    "* Obtain Reference Genomic Data\n",
    "    1. Obtain a reference genome:\n",
    "        * Ensembl, http://ensembl.org\n",
    "        * UCSC, http://genome.ucsc.edu\n",
    "    2. Convert the reference genome into fast, searchable data structure\n",
    "    3. If processing RNA-seq, obtain the gene models\n",
    "    \n",
    "    * GTF/GFF files (Gene Feature Format)\n",
    "    * SAM files (Sequence Alignment/Map format)\n",
    "        * Standardized format for short-read sequence alignments\n",
    "        * Supported by all major alignment packages.\n",
    "        * Cosists of a header section and an alignment section\n",
    "        * Flag section tells us attibutes about the read -> written in binary\n",
    "        * CIGAR section tells us about mapping to reference\n",
    "    * BAM (binary version of SAM)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
